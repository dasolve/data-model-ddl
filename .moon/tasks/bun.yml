fileGroups:
  sources:
    - package.json
    - src/**/*
    - "!**/*.snap.png"
    - "**/*.json"
    - "**/*.ts"
    - "**/*.tsx"
    - "**/*.css"
    - "**/*.sql"
  eslint:
    - /packages/library-eslint/eslint.config.mjs
    - /.oxlintrc.json
    - /tsconfig.options.json
  not-storybook-files:
    - "!**/*.stories.{js,jsx,mjs,ts,tsx}"
    - "!.storybook/**"
  test-files:
    - "**/*.test.{js,jsx,mjs,ts,tsx}"
    - "**/mock.{ts,tsx}"
  not-test-files:
    - "!**/*.test.{js,jsx,mjs,ts,tsx}"
    - "!**/mock.{ts,tsx}"
  design-system:
    - "/packages/library-design-system/**/*.{ts,tsx}"
    - "!/packages/library-design-system/.storybook/**/*"
  ui-components:
    - "/packages/library-ui-components/**/*.{ts,tsx}"
    - "!/packages/library-ui-components/.storybook/**/*"
tasks:
  typecheck:
    toolchain: bun
    command: bun --bun tsc --build
    inputs:
      - "@group(sources)"
      - /tsconfig.options.json
      - /tsconfig.json
      - tsconfig.json
    deps:
      - "^:typecheck"
    type: test
    options:
      runInCI: true
      persistent: false
      mutex: lint
      outputStyle: buffer-only-failure
  oxlint:
    toolchain: bun
    command: bun --bun oxlint
    inputs:
      - "@group(sources)"
      - "@group(eslint)"
    type: test
    options:
      runInCI: true
      persistent: false
      internal: true
      cache: true
      outputStyle: buffer-only-failure
  # lint:
  #   toolchain: bun
  #   command: bun --bun eslint . --report-unused-disable-directives --max-warnings=0
  #   inputs:
  #     - "@group(sources)"
  #     - "@group(eslint)"
  #   deps:
  #     - oxlint
  #     - "^:lint"
  #   type: test
  #   options:
  #     cache: true
  #     timeout: 660
  #     runInCI: true
  #     persistent: false
  #     mutex: lint
  #     outputStyle: buffer-only-failure
  test:
    toolchain: bun
    command: bun test --coverage --coverage-reporter=lcov --coverage-dir=../../coverage/$project --reporter=junit --reporter-outfile=junit.xml
    inputs:
      - "@group(sources)"
    outputs:
      - junit.xml
      - /coverage/$project/lcov.info
    type: test
    deps:
      - "^:test"
    options:
      runInCI: true
      persistent: false
      # This is to avoid scenario where errors are not caught by moon
      timeout: 210
      retryCount: 2
      outputStyle: buffer-only-failure
  test-update-snapshots:
    toolchain: bun
    command: bun test --update-snapshots
    type: build
    local: true
    inputs:
      - "@group(test-files)"
    outputs:
      - "**/__snapshots__/*.snap"
    options:
      persistent: false
      runInCI: false
  format:
    toolchain: bun
    command: prettier --config=../../.prettierrc.json --ignore-path=../../.prettierignore --write .
    inputs:
      - "@group(sources)"
      - /.prettierrc.json
      - /.prettierignore
    type: run
    local: true
    options:
      persistent: false
  format-check:
    toolchain: bun
    command: prettier --config=../../.prettierrc.json --ignore-path=../../.prettierignore --check .
    inputs:
      - "@group(sources)"
      - /.prettierrc.json
      - /.prettierignore
    type: test
    options:
      runInCI: true
      persistent: false
      outputStyle: buffer-only-failure
