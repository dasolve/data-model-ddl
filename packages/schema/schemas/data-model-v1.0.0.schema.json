{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "version": {
      "description": "Data Model DDL file version",
      "type": "string",
      "const": "1.0.0"
    },
    "name": {
      "title": "Data Model Name",
      "type": "string",
      "minLength": 2,
      "maxLength": 59,
      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
    },
    "description": {
      "title": "Data Model Description",
      "description": "Can be used to describe the purpose of the data model",
      "type": "string",
      "maxLength": 255
    },
    "tables": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "title": "Table Name",
            "type": "string",
            "minLength": 2,
            "maxLength": 59,
            "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
          },
          "description": {
            "title": "Table Description",
            "description": "Can be used to describe the purpose of the table in the data model",
            "type": "string",
            "maxLength": 255
          },
          "columns": {
            "minItems": 1,
            "maxItems": 1600,
            "type": "array",
            "items": {
              "anyOf": [
                {
                  "title": "UUID identifier column",
                  "description": "Used for identifying a record uniquely in the table with UUID v4",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "uuid"
                    },
                    "default": {
                      "type": "string",
                      "const": "gen_random_uuid()"
                    },
                    "primary_key": {
                      "default": true,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "default",
                    "primary_key"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Integer identifier column",
                  "description": "Used for identifying a record uniquely in the table with auto-incrementing integer",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "integer"
                    },
                    "default": {
                      "anyOf": [
                        {
                          "type": "string"
                        },
                        {
                          "type": "number"
                        },
                        {
                          "type": "boolean"
                        }
                      ]
                    },
                    "primary_key": {
                      "default": true,
                      "type": "boolean"
                    },
                    "auto_increment": {
                      "default": true,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "primary_key",
                    "auto_increment"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Integer column",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "integer"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "type": "integer",
                      "minimum": -9007199254740991,
                      "maximum": 9007199254740991
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "nullable"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Text column",
                  "description": "Used for storing text content regardless of size.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "text"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "type": "string"
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Date column without time",
                  "description": "Used for storing date values without time.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "date"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "type": "string",
                      "format": "date",
                      "pattern": "^(?:(?:\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-(?:(?:0[13578]|1[02])-(?:0[1-9]|[12]\\d|3[01])|(?:0[469]|11)-(?:0[1-9]|[12]\\d|30)|(?:02)-(?:0[1-9]|1\\d|2[0-8])))$"
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Date column with time",
                  "description": "Used for storing date and time values.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "datetime"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "anyOf": [
                        {
                          "type": "string",
                          "format": "date-time",
                          "pattern": "^(?:(?:\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-(?:(?:0[13578]|1[02])-(?:0[1-9]|[12]\\d|3[01])|(?:0[469]|11)-(?:0[1-9]|[12]\\d|30)|(?:02)-(?:0[1-9]|1\\d|2[0-8])))T(?:(?:[01]\\d|2[0-3]):[0-5]\\d(?::[0-5]\\d(?:\\.\\d+)?)?(?:Z))$"
                        },
                        {
                          "type": "string",
                          "const": "CURRENT_TIMESTAMP()"
                        }
                      ]
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "UUID foreign key column",
                  "description": "Used for identifying a row in another table with UUID v4. It references a column in another table.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "uuid"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "foreign_key": {
                      "type": "object",
                      "properties": {
                        "table": {
                          "type": "string",
                          "minLength": 2,
                          "maxLength": 59,
                          "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                        },
                        "column": {
                          "type": "string",
                          "minLength": 2,
                          "maxLength": 59,
                          "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                        }
                      },
                      "required": [
                        "table",
                        "column"
                      ],
                      "additionalProperties": false
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "foreign_key"
                  ],
                  "additionalProperties": false
                }
              ]
            }
          }
        },
        "required": [
          "name",
          "columns"
        ],
        "additionalProperties": false
      }
    }
  },
  "required": [
    "version",
    "name",
    "tables"
  ],
  "additionalProperties": false
}