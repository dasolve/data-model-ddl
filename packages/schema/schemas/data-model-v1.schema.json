{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "version": {
      "description": "Data Model DDL file version 1.0.0",
      "type": "string",
      "const": "1"
    },
    "name": {
      "title": "Data Model Name",
      "type": "string",
      "minLength": 2,
      "maxLength": 59,
      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
    },
    "dialect": {
      "title": "Database Dialect",
      "description": "The database dialect for which the data model is intended",
      "default": "postgres",
      "type": "string",
      "enum": [
        "postgres"
      ]
    },
    "description": {
      "title": "Data Model Description",
      "description": "Can be used to describe the purpose of the data model",
      "type": "string",
      "maxLength": 255
    },
    "tables": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "title": "Table Name",
            "type": "string",
            "minLength": 2,
            "maxLength": 59,
            "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
          },
          "description": {
            "title": "Table Description",
            "description": "Can be used to describe the purpose of the table in the data model",
            "type": "string",
            "maxLength": 255
          },
          "columns": {
            "minItems": 1,
            "maxItems": 1600,
            "type": "array",
            "items": {
              "anyOf": [
                {
                  "title": "UUID identifier column",
                  "description": "Used for identifying a record uniquely in the table with UUID v4",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "uuid"
                    },
                    "default": {
                      "type": "string",
                      "const": "gen_random_uuid()"
                    },
                    "primary_key": {
                      "default": true,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "default",
                    "primary_key"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Integer identifier column",
                  "description": "Used for identifying a record uniquely in the table with auto-incrementing integer",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "integer"
                    },
                    "generated_always_as": {
                      "type": "string",
                      "const": "identity"
                    },
                    "primary_key": {
                      "default": true,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "generated_always_as",
                    "primary_key"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Integer foreign key column",
                  "description": "Used for identifying a row in another table with an integer. It references a column in another table.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "integer"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "foreign_key": {
                      "type": "object",
                      "properties": {
                        "table": {
                          "type": "string",
                          "minLength": 2,
                          "maxLength": 59,
                          "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                        },
                        "column": {
                          "type": "string",
                          "minLength": 2,
                          "maxLength": 59,
                          "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                        }
                      },
                      "required": [
                        "table",
                        "column"
                      ],
                      "additionalProperties": false
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "foreign_key"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Decimal number column",
                  "description": "Used for storing decimal numbers with a fixed precision.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "object",
                      "properties": {
                        "name": {
                          "type": "string",
                          "const": "numeric"
                        },
                        "options": {
                          "type": "object",
                          "properties": {
                            "precision": {
                              "type": "number",
                              "minimum": 0,
                              "maximum": 1000
                            },
                            "scale": {
                              "type": "number",
                              "minimum": 0,
                              "maximum": 1000
                            }
                          },
                          "required": [
                            "precision",
                            "scale"
                          ],
                          "additionalProperties": false
                        }
                      },
                      "required": [
                        "name",
                        "options"
                      ],
                      "additionalProperties": false
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "anyOf": [
                        {
                          "type": "string"
                        },
                        {
                          "type": "number"
                        }
                      ]
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    },
                    "generated_always_as": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "name",
                    "type"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Integer column",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "integer"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "type": "integer",
                      "minimum": -9007199254740991,
                      "maximum": 9007199254740991
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    },
                    "generated_always_as": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "nullable"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Text column",
                  "description": "Used for storing text content regardless of size.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "text"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "type": "string"
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    },
                    "generated_always_as": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "name",
                    "type"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Date column without time",
                  "description": "Used for storing date values without time.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "date"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "type": "string",
                      "const": "CURRENT_DATE()"
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "Date column with time",
                  "description": "Used for storing date and time values.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "object",
                      "properties": {
                        "name": {
                          "type": "string",
                          "const": "timestamp"
                        },
                        "options": {
                          "type": "object",
                          "properties": {
                            "withTimeZone": {
                              "default": true,
                              "type": "boolean"
                            }
                          },
                          "required": [
                            "withTimeZone"
                          ],
                          "additionalProperties": false
                        }
                      },
                      "required": [
                        "name",
                        "options"
                      ],
                      "additionalProperties": false
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "default": {
                      "type": "string",
                      "const": "now()"
                    },
                    "unique": {
                      "default": false,
                      "type": "boolean"
                    }
                  },
                  "required": [
                    "name",
                    "type"
                  ],
                  "additionalProperties": false
                },
                {
                  "title": "UUID foreign key column",
                  "description": "Used for identifying a row in another table with UUID v4. It references a column in another table.",
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string",
                      "minLength": 2,
                      "maxLength": 59,
                      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                    },
                    "description": {
                      "type": "string",
                      "maxLength": 255
                    },
                    "type": {
                      "type": "string",
                      "const": "uuid"
                    },
                    "nullable": {
                      "default": false,
                      "type": "boolean"
                    },
                    "foreign_key": {
                      "type": "object",
                      "properties": {
                        "table": {
                          "type": "string",
                          "minLength": 2,
                          "maxLength": 59,
                          "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                        },
                        "column": {
                          "type": "string",
                          "minLength": 2,
                          "maxLength": 59,
                          "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
                        }
                      },
                      "required": [
                        "table",
                        "column"
                      ],
                      "additionalProperties": false
                    }
                  },
                  "required": [
                    "name",
                    "type",
                    "foreign_key"
                  ],
                  "additionalProperties": false
                }
              ]
            }
          }
        },
        "required": [
          "name",
          "columns"
        ],
        "additionalProperties": false
      }
    }
  },
  "required": [
    "version",
    "name",
    "tables"
  ],
  "additionalProperties": false
}