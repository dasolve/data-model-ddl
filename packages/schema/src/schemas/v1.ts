import { z } from "zod/v4";
import {
  integerPrimaryKeyColumn,
  uuidPrimaryKeyColumn,
  integerForeignKeyColumn,
  uuidForeignKeyColumn,
} from "../columns/keys";
import { decimalNumberColumn, integerColumn } from "../columns/numbers";
import { textColumn } from "../columns/text";
import { dateColumn, datetimeColumn } from "../columns/time";
import { tableColumn } from "../columns/common";

export const schema = z.strictObject({
  version: z.literal("1").describe("Data Model DDL file version 1.0.0"),
  name: tableColumn.shape.name.meta({ title: "Data Model Name" }),
  dialect: z.enum(["postgres"]).default("postgres").optional().meta({
    title: "Database Dialect",
    description: "The database dialect for which the data model is intended",
  }),
  description: tableColumn.shape.description.meta({
    title: "Data Model Description",
    description: "Can be used to describe the purpose of the data model",
  }),
  tables: z.array(
    z.strictObject({
      name: tableColumn.shape.name.meta({
        title: "Table Name",
      }),
      description: tableColumn.shape.description.meta({
        title: "Table Description",
        description:
          "Can be used to describe the purpose of the table in the data model",
      }),
      columns: z
        .array(
          z.union([
            uuidPrimaryKeyColumn,
            integerPrimaryKeyColumn,
            integerForeignKeyColumn,
            decimalNumberColumn,
            integerColumn,
            textColumn,
            dateColumn,
            datetimeColumn,
            uuidForeignKeyColumn,
          ])
        )
        .nonempty()
        .max(1600),
    })
  ),
});
