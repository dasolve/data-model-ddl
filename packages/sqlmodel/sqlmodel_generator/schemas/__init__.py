"""
Schema definitions for validating data model files.
"""
