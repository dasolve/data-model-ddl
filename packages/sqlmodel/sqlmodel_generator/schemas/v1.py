"""
Schema definition for v1 of the data model.
"""

from typing import Dict, List, Optional, Union

from pydantic import BaseModel, Field


class ForeignKey(BaseModel):
    """Foreign key definition."""

    table: str
    column: str


class ColumnTypeOptions(BaseModel):
    """Options for complex column types."""

    precision: Optional[int] = None
    scale: Optional[int] = None
    withTimeZone: Optional[bool] = None


class ColumnType(BaseModel):
    """Complex column type with options."""

    name: str
    options: Optional[ColumnTypeOptions] = None


class BaseColumn(BaseModel):
    """Base column definition with common properties."""

    name: str
    description: Optional[str] = None
    nullable: bool = False
    unique: bool = False
    generated_always_as: Optional[str] = None


class UUIDColumn(BaseColumn):
    """UUID column definition."""

    type: str
    default: Optional[str] = None
    primary_key: bool = False
    foreign_key: Optional[ForeignKey] = None


class IntegerColumn(BaseColumn):
    """Integer column definition."""

    type: str
    default: Optional[int] = None
    primary_key: bool = False
    generated_always_as: Optional[str] = None
    foreign_key: Optional[ForeignKey] = None


class NumericColumn(BaseColumn):
    """Numeric column definition."""

    type: ColumnType
    default: Optional[Union[str, float]] = None


class TextColumn(BaseColumn):
    """Text column definition."""

    type: str
    default: Optional[str] = None


class DateColumn(BaseColumn):
    """Date column definition."""

    type: str
    default: Optional[str] = None


class TimestampColumn(BaseColumn):
    """Timestamp column definition."""

    type: ColumnType
    default: Optional[str] = None


# Union of all column types for type annotation
ColumnTypes = Union[
    UUIDColumn, IntegerColumn, NumericColumn, TextColumn, DateColumn, TimestampColumn
]


class Table(BaseModel):
    """Table definition."""

    name: str
    description: Optional[str] = None
    columns: List[ColumnTypes]


class DataModel(BaseModel):
    """Root data model definition."""

    version: str
    name: str
    dialect: str = "postgres"
    description: Optional[str] = None
    tables: List[Table]


# The schema as a variable that can be imported
schema = DataModel
