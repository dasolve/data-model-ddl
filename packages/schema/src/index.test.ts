import { describe, expect, it } from "bun:test";
import schemas from "./index";

describe("Schema Tests", () => {
  it("should export the schema", () => {
    expect(schemas).toMatchSnapshot();
  });
});
