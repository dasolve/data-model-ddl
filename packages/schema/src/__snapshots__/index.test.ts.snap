// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Tests should export the schema 1`] = `
{
  "1": _ {
    "_def": {
      "catchall": _ {
        "_def": {
          "type": "never",
        },
        "_zod": {
          "bag": {},
          "constr": [Function: _],
          "def": {
            "type": "never",
          },
          "deferred": [
            [Function],
            [Function],
          ],
          "id": [native code],
          "parse": [Function],
          "run": [Function],
          "traits":           
Set {
            "ZodNever",
            "$ZodNever",
            "$ZodType",
            "ZodType",
          }
,
          "version": {
            "major": 4,
            "minor": 0,
            "patch": 0,
          },
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "check": [Function],
        "clone": [Function],
        "def": {
          "type": "never",
        },
        "default": [Function],
        "describe": [Function],
        "description": [native code],
        "isNullable": [Function],
        "isOptional": [Function],
        "meta": [Function],
        "nonoptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "overwrite": [Function],
        "parse": [Function],
        "parseAsync": [Function: AsyncFunction],
        "pipe": [Function],
        "prefault": [Function],
        "readonly": [Function],
        "refine": [Function],
        "register": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function: AsyncFunction],
        "spa": [Function: AsyncFunction],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [native code],
      "type": "object",
    },
    "_zod": {
      "bag": {},
      "constr": [Function: _],
      "def": {
        "catchall": _ {
          "_def": {
            "type": "never",
          },
          "_zod": {
            "bag": {},
            "constr": [Function: _],
            "def": {
              "type": "never",
            },
            "deferred": [
              [Function],
              [Function],
            ],
            "id": [native code],
            "parse": [Function],
            "run": [Function],
            "traits":             
Set {
              "ZodNever",
              "$ZodNever",
              "$ZodType",
              "ZodType",
            }
,
            "version": {
              "major": 4,
              "minor": 0,
              "patch": 0,
            },
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "check": [Function],
          "clone": [Function],
          "def": {
            "type": "never",
          },
          "default": [Function],
          "describe": [Function],
          "description": [native code],
          "isNullable": [Function],
          "isOptional": [Function],
          "meta": [Function],
          "nonoptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "overwrite": [Function],
          "parse": [Function],
          "parseAsync": [Function: AsyncFunction],
          "pipe": [Function],
          "prefault": [Function],
          "readonly": [Function],
          "refine": [Function],
          "register": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function: AsyncFunction],
          "spa": [Function: AsyncFunction],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "shape": [native code],
        "type": "object",
      },
      "deferred": [
        [Function],
        [Function],
      ],
      "id": [native code],
      "parse": [Function],
      "propValues": [native code],
      "run": [Function],
      "traits":       
Set {
        "ZodObject",
        "$ZodObject",
        "$ZodType",
        "ZodType",
      }
,
      "version": {
        "major": 4,
        "minor": 0,
        "patch": 0,
      },
    },
    "and": [Function],
    "array": [Function],
    "brand": [Function],
    "catch": [Function],
    "catchall": [Function],
    "check": [Function],
    "clone": [Function],
    "def": {
      "catchall": _ {
        "_def": {
          "type": "never",
        },
        "_zod": {
          "bag": {},
          "constr": [Function: _],
          "def": {
            "type": "never",
          },
          "deferred": [
            [Function],
            [Function],
          ],
          "id": [native code],
          "parse": [Function],
          "run": [Function],
          "traits":           
Set {
            "ZodNever",
            "$ZodNever",
            "$ZodType",
            "ZodType",
          }
,
          "version": {
            "major": 4,
            "minor": 0,
            "patch": 0,
          },
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "check": [Function],
        "clone": [Function],
        "def": {
          "type": "never",
        },
        "default": [Function],
        "describe": [Function],
        "description": [native code],
        "isNullable": [Function],
        "isOptional": [Function],
        "meta": [Function],
        "nonoptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "overwrite": [Function],
        "parse": [Function],
        "parseAsync": [Function: AsyncFunction],
        "pipe": [Function],
        "prefault": [Function],
        "readonly": [Function],
        "refine": [Function],
        "register": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function: AsyncFunction],
        "spa": [Function: AsyncFunction],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [native code],
      "type": "object",
    },
    "default": [Function],
    "describe": [Function],
    "description": [native code],
    "extend": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "keyof": [Function],
    "loose": [Function],
    "merge": [Function],
    "meta": [Function],
    "nonoptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "omit": [Function],
    "optional": [Function],
    "or": [Function],
    "overwrite": [Function],
    "parse": [Function],
    "parseAsync": [Function: AsyncFunction],
    "partial": [Function],
    "passthrough": [Function],
    "pick": [Function],
    "pipe": [Function],
    "prefault": [Function],
    "readonly": [Function],
    "refine": [Function],
    "register": [Function],
    "required": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function: AsyncFunction],
    "shape": [native code],
    "spa": [Function: AsyncFunction],
    "strict": [Function],
    "strip": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
}
`;
